sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala

```diff
@@ -496,19 +511,38 @@ case class Overlay(input: Expression, replace: Expression, pos: Expression, len:
     this(str, replace, pos, Literal.create(-1, IntegerType))
   }
 
-  override def dataType: DataType = StringType
+  override def dataType: DataType = input.dataType
 
-  override def inputTypes: Seq[AbstractDataType] =
-    Seq(StringType, StringType, IntegerType, IntegerType)
+  override def inputTypes: Seq[AbstractDataType] = Seq(TypeCollection(StringType, BinaryType),
+    TypeCollection(StringType, BinaryType), IntegerType, IntegerType)
 
   override def children: Seq[Expression] = input :: replace :: pos :: len :: Nil
 
+  override def checkInputDataTypes(): TypeCheckResult = {
+    val (inputType, replaceType, posType, lenType) =
+      (input.dataType, replace.dataType, pos.dataType, len.dataType)
+    (inputType, replaceType) match {
+      case (StringType, StringType) | (BinaryType, BinaryType) =>
```

Comment:
Is this enough just to check `inputType.sameType(replaceType)`?

Reply:
If we only plan to implement `overlay(string, string)`, `overlay(binary, binary)`, this is enough.

Reply:
```
  override def checkInputDataTypes(): TypeCheckResult = {
    TypeUtils.checkForSameTypeInputExpr(
      input.dataType :: replace.dataType :: Nil, s"function $prettyName")
  }
```

Reply:
OK. This is a good idea.

Reply:
I forgot add the default type check;
```

  override def checkInputDataTypes(): TypeCheckResult = {
    val inputTypeCheck = super.checkInputDataTypes()
    if (inputTypeCheck.isSuccess) {
      TypeUtils.checkForSameTypeInputExpr(
        input.dataType :: replace.dataType :: Nil, s"function $prettyName")
    } else {
      inputTypeCheck
    }
  }
```
I think overlay(int, int) and overlay(long, long) are valid queries for overlay because their types can be implictly casted to strings.

Reply:
I forgot add the default type check;
```

  override def checkInputDataTypes(): TypeCheckResult = {
    val inputTypeCheck = super.checkInputDataTypes()
    if (inputTypeCheck.isSuccess) {
      TypeUtils.checkForSameTypeInputExpr(
        input.dataType :: replace.dataType :: Nil, s"function $prettyName")
    } else {
      inputTypeCheck
    }
  }
```
I think overlay(int, int) and overlay(long, long) are valid queries for overlay because their types can be implictly casted to strings.

Reply:
Thanks, this is OK. I accept your suggestion.